src/impl_methods.rs

```diff
@@ -630,11 +629,18 @@ where
         S: DataMut,
         I: NdIndex<D>,
     {
-        let ptr1: *mut _ = &mut self[index1];
-        let ptr2: *mut _ = &mut self[index2];
-        unsafe {
-            std_ptr::swap(ptr1, ptr2);
+        let ptr = self.as_mut_ptr();
+        let offset1 = index1.index_checked(&self.dim, &self.strides);
+        let offset2 = index2.index_checked(&self.dim, &self.strides);
+        if let Some(offset1) = offset1 {
```

Comment:
Fwiw, as of Rust 1.46, things like this can be written a little more nicely using [`Option::zip`](https://doc.rust-lang.org/std/option/enum.Option.html#method.zip):

```rust
if let Some((offset1, offset2)) = offset1.zip(offset2) {
    unsafe { ... }
} else {
    panic!(...);
}
```

(or `if let Some((offset1, offset2)) = Option::zip(offset1, offset2)` if that looks better)

Reply:
if let (Some(_), Some(_)) = (offset1, offset2) also works IMO. Now which option gives better code :) I wouldn't bother to investigate though.